### PR TITLE
Rewrite MQTT PUBLISH topic parser for C2D / IoT Edge parsing

### PR DESCRIPTION
**Pull request by jspaith**
The code that parses MQTT PUBLISH for C2D/IoT Edge needed some work.  Specifically it was dropping property values for Edge and further investigation shows there were (obscure) error cases where a leak is possible.  It was also getting hard to follow with the various features added over the years.

Rather than patch, this is a rewrite.  Gets rid of the complicated property detection logic (and the fact it was duplicated) as well as input queue detection based on the topic name and not the extraneous tokenization.

Finally this adds an integration test for this - `iothubtransport_mqtt_common_int_ut`.  Being able to test all the MQTT TOPIC permutations with existing UT framework would've been very burdensome.  This only mocks the absolute minimum required (namely umqtt layer plumbing up topics) and most importantly does NOT mock anything in c-utility or the IoTHub_Message layer.

### CODE REVIEW

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 1]
```diff
@@ -1263,357 +1312,434 @@ static int subscribeToNotifyStateIfNeeded(PMQTTTRANSPORT_HANDLE_DATA transport_d
...
+    if (type == IOTHUB_TYPE_EVENT_QUEUE)
+    {
+        const char* input_queue_topic = STRING_c_str(transportData->topic_InputQueue);
+        // Substract one from queue length to reflect this having an extra # we subscribed to
+        propertiesStart = topic_name + (strlen(input_queue_topic) - 1);
+    }
+    else if (type == IOTHUB_TYPE_TELEMETRY)
+    {
+        const char* mqtt_messsage_queue_topic = STRING_c_str(transportData->topic_MqttMessage);
+        propertiesStart = topic_name + (strlen(mqtt_messsage_queue_topic) - 1);
```
Comment (danewalton) [inline iothub_client/src/iothubtransport_mqtt_common.c:1545]: might want to add the comment here too about the subtract one for the `#`

**Review by danewalton — COMMENTED** [review]
just some initial comments for now

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 2]
```diff
@@ -107,30 +109,64 @@ static const char DEFAULT_IOTHUB_PRODUCT_IDENTIFIER[] = CLIENT_DEVICE_TYPE_PREFI
...
+    { DEFINE_MQTT_SYSTEM_PROPERTY(MESSAGE_ID_PROPERTY), MQTT_PROPERTY_TYPE_MESSAGE_ID},
+    { DEFINE_MQTT_SYSTEM_PROPERTY(MESSAGE_USER_ID), MQTT_PROPERTY_TYPE_MESSAGE_USER_ID },
+    { DEFINE_MQTT_SYSTEM_PROPERTY(CORRELATION_ID_PROPERTY), MQTT_PROPERTY_TYPE_CORRELATION_ID },
+    { DEFINE_MQTT_SYSTEM_PROPERTY(CONTENT_TYPE_PROPERTY), MQTT_PROPERTY_TYPE_CONTENT_TYPE },
+    { DEFINE_MQTT_SYSTEM_PROPERTY(CONTENT_ENCODING_PROPERTY), MQTT_PROPERTY_TYPE_CONTENT_ENCODING },
+    { DEFINE_MQTT_SYSTEM_PROPERTY(CONNECTION_DEVICE_ID), MQTT_PROPERTY_TYPE_CONNECTION_DEVICE_ID},
+    { DEFINE_MQTT_SYSTEM_PROPERTY(CONNECTION_MODULE_ID_PROPERTY), MQTT_PROPERTY_TYPE_CONNECTION_MODULE_ID },
+    // "System" properties the SDK previously ignored and will continue to do so for compat.
+    { DEFINE_MQTT_SYSTEM_PROPERTY(MESSAGE_CREATION_TIME_UTC), MQTT_PROPERTY_TYPE_CREATION_TIME},
+    { DEFINE_MQTT_SYSTEM_PROPERTY("on"), MQTT_PROPERTY_TYPE_SILENTLY_IGNORE },
```
[the remark this answers is not in the document]
Comment (jspaith) [reply iothub_client/src/iothubtransport_mqtt_common.c]: Nope, oversight.  Fixing and looking at naming it's really all over the place - some end in PROPERTY but others don't, e.g.  Moving so everythiing starts with SYS_PROP_ prefix so intent is a bit cleaner.

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 3]
```diff
@@ -576,28 +612,41 @@ static int parseDeviceTwinTopicInfo(const char* resp_topic, bool* patch_msg, siz
...
+        result = 0;
     }
     else if (InternStrnicmp(topic_resp, TOPIC_DEVICE_METHOD_PREFIX, sizeof(TOPIC_DEVICE_METHOD_PREFIX) - 1) == 0)
     {
-        type = IOTHUB_TYPE_DEVICE_METHODS;
+        *type = IOTHUB_TYPE_DEVICE_METHODS;
+        result = 0;
+    }
+    // mqtt_messsage_queue_topic contains additional "#" from subscribe, which we strip off on comparing incoming.
+    else if (((mqtt_messsage_queue_topic = STRING_c_str(transportData->topic_MqttMessage)) != NULL) && (InternStrnicmp(topic_resp, mqtt_messsage_queue_topic, strlen(mqtt_messsage_queue_topic) - 1) == 0))
```
[the remark this answers is not in the document]
Comment (jspaith) [reply iothub_client/src/iothubtransport_mqtt_common.c]: Fixing.

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 4]
```diff
@@ -1263,357 +1312,434 @@ static int subscribeToNotifyStateIfNeeded(PMQTTTRANSPORT_HANDLE_DATA transport_d
 }
 
 //
-// isSystemProperty returns whether a given property name in an MQTT TOPIC published to this device/module
-// is considered a "system".  MQTT does not have a protocol defined concept of system properties.  In this usage
-// it implies that the IOTHUB_MESSAGE_HANDLE has an API for direct manipulation of the property (e.g. IoTHubMessage_GetMessageId).
+// GetMqttPropertyType compares a specific property in a topic to whetehr its one that we know about
```
[the remark this answers is not in the document]
Comment (jspaith) [reply iothub_client/src/iothubtransport_mqtt_common.c]: fixed.

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 5]
```diff
@@ -1263,357 +1312,434 @@ static int subscribeToNotifyStateIfNeeded(PMQTTTRANSPORT_HANDLE_DATA transport_d
...
+// (specifically whether IOTHUB_MESSAGE_HANDLE has an accessor for it).
 //
-static bool isSystemProperty(const char* tokenData)
+static MQTT_PROPERTY_TYPE GetMqttPropertyType(const char* propertyNameAndValue, size_t propertyNameLength)
```
[the remark this answers is not in the document]
Comment (jspaith) [reply iothub_client/src/iothubtransport_mqtt_common.c:1321]: It's indeed `propertyNameAndValue`.  The way that the tokenizer works in `extractMqttProperties` is it tokenizes off of the "&", and we don't know whether we actually need to stash off the name part *until* we perform this check.  So if it's like contentType or whatever we just convert to the enum and if it's app property then we split out

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 6]
```diff
@@ -1263,357 +1312,434 @@ static int subscribeToNotifyStateIfNeeded(PMQTTTRANSPORT_HANDLE_DATA transport_d
...
 
 //
-// addInputNamePropertyToMsg translates the input name (embedded in the MQTT topic name) into the IoTHubMessage handle
+// addInputNamePropertyToMsg translates the input name (embedded in the MQTT topic name) into the iotHubMessage handle
 // such that the application can call IoTHubMessage_GetInputName() to retrieve this.  This is only currently used 
 // in IoT Edge module to module message communication, so that this module receiving the message can know which module invoked in.
 //
-static int addInputNamePropertyToMsg(IOTHUB_MESSAGE_HANDLE IoTHubMessage, const char* topic_name)
+// For IoT Edge module to module communication, the incoming topic will be of the form devices/{deviceId}/modules/{moduleId}/inputs/{inputName}.
+// When this function is called, the caller has already skipped past the topic_InputQueue.  In example, we would start at inputs/{inputName}.
```
[the remark this answers is not in the document]
Comment (jspaith) [reply iothub_client/src/iothubtransport_mqtt_common.c]: It was referring to the variable name `topic_InputQueue` which is part of `PMQTTTRANSPORT_HANDLE_DATA`.  But that's not helpful; cleanin up to reference more.

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 7]
```diff
@@ -1263,357 +1312,434 @@ static int subscribeToNotifyStateIfNeeded(PMQTTTRANSPORT_HANDLE_DATA transport_d
...
-    int result = MU_FAILURE;
-    int number_tokens_read = 0;
+    const char* result;
+    const char* inputNameStart;
+    const char* inputNameEnd;
+    char* inputNameCopy = NULL;
 
-    STRING_TOKENIZER_HANDLE token_handle = STRING_TOKENIZER_create_from_char(topic_name);
-    if (token_handle == NULL)
+    if (((inputNameStart = strchr(propertiesStart, TOPIC_SLASH)) == NULL) || (*(inputNameStart+1) == 0))
```
[the remark this answers is not in the document]
Comment (jspaith) [reply iothub_client/src/iothubtransport_mqtt_common.c]: Actually - I think technically it should be '\0'.  NULL is a pointer and could be (void*)0 whereas `\0` is a char.

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 8]
```diff
@@ -1263,357 +1312,434 @@ static int subscribeToNotifyStateIfNeeded(PMQTTTRANSPORT_HANDLE_DATA transport_d
...
+            {
+                LogError("Cannot allocate input name");
+                result = NULL;
+            }
+            else
             {
-                number_tokens_read++;
-                if (number_tokens_read == (slashes_to_reach_input_name + 1))
+                memcpy(inputNameCopy, inputNameStart, inputNameLength);
+                inputNameCopy[inputNameLength] = 0;
```
[the remark this answers is not in the document]
Comment (jspaith) [reply iothub_client/src/iothubtransport_mqtt_common.c]: `'\0'`

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 9]
```diff
@@ -1263,357 +1312,434 @@ static int subscribeToNotifyStateIfNeeded(PMQTTTRANSPORT_HANDLE_DATA transport_d
...
+
+                if (IoTHubMessage_SetInputName(iotHubMessage, inputNameCopy) != IOTHUB_MESSAGE_OK)
```
[the remark this answers is not in the document]
Comment (jspaith) [reply iothub_client/src/iothubtransport_mqtt_common.c:1381]: It does.  I can't change `IoTHubMessage_SetInputName` since it's public; if we cared strongly enough we could add private helpers to message functions that could transfer pointer ownership if it could trust caller was handing it off.

Don't think it's worth it given how infrequently this code gets run and that we don't necessarily have patterns like this in other places of message or much generally.

```c
IOTHUB_MESSAGE_RESULT IoTHubMessage_SetInputName(IOTHUB_MESSAGE_HANDLE iotHubMessageHandle, const char* inputName)
...
if (mallocAndStrcpy_s(&handleData->inputName, inputName) != 0)
{
...
```

File: iothub_client/src/iothubtransport_mqtt_common.c (hunk 1 above)
Comment (jspaith) [reply iothub_client/src/iothubtransport_mqtt_common.c]: fixed.

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c [hunk 10]
```diff
@@ -0,0 +1,1172 @@
...
+static const size_t mqttNoMatchTopicLength = sizeof(mqttNoMatchTopic) / sizeof(mqttNoMatchTopic[0]);
+
+static const char* emptyPropertyMQTTTopics[] = {
+    "devices/myDeviceId/messages/devicebound/",
+    "devices/myDeviceId/messages/devicebound/&",
+    "devices/myDeviceId/messages/devicebound/&&",
+    "devices/myDeviceId/messages/devicebound/&&&",
+    "devices/myDeviceId/messages/devicebound/=",
+    "devices/myDeviceId/messages/devicebound/fooBar",
+    //"devices/myDeviceId/messages/devicebound/==",
```
Comment (danewalton) [inline iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c:283]: needed?

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c [hunk 11]
```diff
@@ -0,0 +1,1172 @@
...
+};
+
+static const size_t emptyMQTTTopicsLength = sizeof(emptyPropertyMQTTTopics) / sizeof(emptyPropertyMQTTTopics[0]);  
+TEST_EXPECTED_MESSAGE_PROPERTIES noProperties = { NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL};
+
+static const char* TEST_MQTT_MESSAGE_APP_PROPERTIES_1 = "devices/myDeviceId/messages/devicebound/customKey1=customValue1";
```
Comment (danewalton) [inline iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c:289]: remind me we don't have `&` after that last `/` and before properties? or is that somewhere else?

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c [hunk 12]
```diff
@@ -0,0 +1,1172 @@
...
+char* expectedValue2[] = {"customValue1", "customValue2"};
+TEST_EXPECTED_APPLICATION_PROPERTIES app2 = { expectedKey2, expectedValue2, 2};
+TEST_EXPECTED_MESSAGE_PROPERTIES expectedAppProperties2 = { NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, &app2};
+
+#define TEST_APP_PROPERTY_KEY1 "temperature%2FAlert1"
+#define TEST_APP_PROPERTY_KEY2 "temperature%2FAlert2"
+#define TEST_APP_PROPERTY_KEY3 "temperature%2FAlert3"
+
+#define TEST_APP_PROPERTY_VALUE1 "false%251"
+#define TEST_APP_PROPERTY_VALUE2 "false%252"
```
Comment (danewalton) [inline iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c:306]: should these be 
```suggestion
#define TEST_APP_PROPERTY_VALUE1 "false1%25"
#define TEST_APP_PROPERTY_VALUE2 "false2%25"
```
or was this done on purpose (the next one is this way)

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c [hunk 13]
```diff
@@ -0,0 +1,1172 @@
...
+TEST_EXPECTED_MESSAGE_PROPERTIES allInputSystemPropertiesManyApplicationSet1 = { TEST_CONTENT_TYPE_VALUE, TEST_CONTENT_ENCODING_VALUE, TEST_MSG_ID_VALUE, TEST_CORRELATION_PROPERTY, TEST_INPUT_QUEUE_1, TEST_CONNECTION_MODULE_VALUE, TEST_CONNECTION_DEVICE_VALUE, TEST_CREATION_TIME_VALUE, TEST_MSG_USER_ID_VALUE, &app3};
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#ifdef __cplusplus
+}
+#endif
```
Comment (danewalton) [inline iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c:386]: needed?

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c [hunk 14]
```diff
@@ -0,0 +1,1172 @@
...
+}
+
+static MQTT_CLIENT_HANDLE my_mqtt_client_init(ON_MQTT_MESSAGE_RECV_CALLBACK msgRecv, ON_MQTT_OPERATION_CALLBACK opCallback, void* callbackCtx, ON_MQTT_ERROR_CALLBACK errorCallback, void* errorcallbackCtx)
+{
+    g_fnMqttMsgRecv = msgRecv;
+    g_fnMqttOperationCallback = opCallback;
+    g_callbackCtx = callbackCtx;
+    g_fnMqttErrorCallback = errorCallback;
+    g_errorcallbackCtx = errorcallbackCtx;
+    return (MQTT_CLIENT_HANDLE)malloc(12);
```
Comment (danewalton) [inline iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c:440]: magic number? How is this calc'd?

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c [hunk 15]
```diff
@@ -0,0 +1,1172 @@
...
+static void my_mqtt_client_deinit(MQTT_CLIENT_HANDLE handle)
+{
+    free(handle);
+}
+
+static XIO_HANDLE my_xio_create(const IO_INTERFACE_DESCRIPTION* io_interface_description, const void* xio_create_parameters)
+{
+    (void)io_interface_description;
+    (void)xio_create_parameters;
+    return (XIO_HANDLE)malloc(1);
```
Comment (danewalton) [inline iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c:460]: same here and elsewhere. Can this just be a `sizeof(XIO_HANDLE)`?

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c [hunk 16]
```diff
@@ -0,0 +1,1172 @@
...
+    ASSERT_ARE_EQUAL(char_ptr, expectedMessageProperties->inputName, inputName, "Input names don't match for topic %s", topicToTest);
+
+    const char* connectionModuleId = IoTHubMessage_GetConnectionModuleId(g_messageFromCallback);
+    ASSERT_ARE_EQUAL(char_ptr, expectedMessageProperties->connectionModuleId, connectionModuleId, "Connection module ids don't match for topic %s", topicToTest);
+
+    const char* connectionDeviceId = IoTHubMessage_GetConnectionDeviceId(g_messageFromCallback);
+    ASSERT_ARE_EQUAL(char_ptr, expectedMessageProperties->connectionDeviceId, connectionDeviceId, "Connection device ids don't match for topic %s", topicToTest);
+
+    const char* messageCreationTime = IoTHubMessage_GetMessageCreationTimeUtcSystemProperty(g_messageFromCallback);
+    ASSERT_ARE_EQUAL(char_ptr, expectedMessageProperties->messageCreationTime, messageCreationTime, "Message creation tims don't match for topic %s", topicToTest);
```
Comment (danewalton) [inline iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c:796]:
```suggestion
    ASSERT_ARE_EQUAL(char_ptr, expectedMessageProperties->messageCreationTime, messageCreationTime, "Message creation times don't match for topic %s", topicToTest);
```

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c [hunk 17]
```diff
@@ -0,0 +1,1172 @@
...
+    // parsed on an MQTT PUBLISH to the device itself and hence in the IoTHubMessage layer they'll always be NULL.
+    ASSERT_IS_NULL(IoTHubMessage_GetOutputName(g_messageFromCallback), "GetOutputName should have returned NULL for topic %s", topicToTest);
+    ASSERT_IS_NULL(IoTHubMessage_GetDiagnosticPropertyData(g_messageFromCallback), "GetDiagnosticPropertyData should have returned NULL for topic %s", topicToTest);
+
+    // Check application properties
+    MAP_HANDLE mapHandle = IoTHubMessage_Properties(g_messageFromCallback);
+    ASSERT_IS_NOT_NULL(mapHandle);
+
+    const char*const* actualKeys;
+    const char*const* actualValues;
```
Comment (danewalton) [inline iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c:811]: nit for these
```suggestion
    const char* const* actualKeys;
    const char* const* actualValues;
```

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c [hunk 18]
```diff
@@ -0,0 +1,1172 @@
...
+    size_t actualKeysLen;
+    size_t expectedKeyLen = (expectedMessageProperties->applicationProperties != NULL) ? expectedMessageProperties->applicationProperties->keysLength : 0;
+
+    ASSERT_ARE_EQUAL(MAP_RESULT, MAP_OK, Map_GetInternals(mapHandle, &actualKeys, &actualValues, &actualKeysLen));
+    ASSERT_ARE_EQUAL(int, expectedKeyLen, actualKeysLen, "Number of custom properties don't match for topic %s", topicToTest);
+
+    for (i = 0; i < expectedKeyLen; i++)
```
Comment (danewalton) [inline iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c:818]: nit: why have `size_t i` defined at the top of the func?

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c [hunk 19]
```diff
@@ -0,0 +1,1172 @@
...
+    {
+        ASSERT_ARE_EQUAL(char_ptr, expectedMessageProperties->applicationProperties->values[i], IoTHubMessage_GetProperty(g_messageFromCallback, expectedMessageProperties->applicationProperties->keys[i]), "Properties don't match for topic %s", topicToTest);
+    }
+}
+
+
+//
+// TestMessageProcessing invokes the MQTT PUBLISH to device callback code, which will (on success) will store
```
Comment (danewalton) [inline iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c:826]:
```suggestion
// TestMessageProcessing invokes the MQTT PUBLISH to device callback code, which will (on success) store
```

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c [hunk 20]
```diff
@@ -0,0 +1,1172 @@
...
+// the parsed message into the test's g_messageFromCallback.  TestMesageProcessing then verifies message is expected.
```
Comment (danewalton) [inline iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c:827]:
```suggestion
// the parsed message into the test's g_messageFromCallback.  TestMessageProcessing then verifies message is expected.
```

**Review by danewalton — APPROVED** [review]

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c (hunk 10 above)
Comment (jspaith) [reply iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c]: It didn't belong there since we end up with a name="" and value="=" on this case, which is really weird.  Mostly test should be there to make sure SDK doesn't crash with all these what turn into 0 len strings thrown at us.  Adding a separate test case for this.

**Review by jspaith — COMMENTED** [review]
Incorporating your feedback & some additional review.

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c (hunk 12 above)
Comment (jspaith) [reply iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c]: fixed.

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 21]
```diff
@@ -1291,329 +1338,409 @@ static bool isSystemProperty(const char* tokenData)
...
         {
-            LogError("Failed allocating token_value");
+            LogError("Cannot find '/' after input name");
+            result = NULL;
         }
-        else
+        else 
         {
-            while (STRING_TOKENIZER_get_next_token(token_handle, token_value, "/") == 0)
+            size_t inputNameLength = inputNameEnd - inputNameStart;
```
Comment (ewertons) [inline iothub_client/src/iothubtransport_mqtt_common.c:1367]: Should we care if Edge - in a super rare situation - sent us a topic with empty input name ("//")? In that case `inputNameLength` would be 0 and we would set an empty input name in the message handle. But that would be wrong... But we would also be just conveying what the service (in this case Edge) is giving us.

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 22]
```diff
@@ -1291,329 +1338,409 @@ static bool isSystemProperty(const char* tokenData)
...
-
-        if (number_tokens_read != (slashes_to_reach_input_name + 1))
-        {
-            LogError("Not enough '/' to contain input name.  Got %d, need at least %d", number_tokens_read, (slashes_to_reach_input_name + 1));
-            result = MU_FAILURE;
-        }
-        STRING_TOKENIZER_destroy(token_handle);
     }
 
+    free(inputNameCopy);
```
Comment (ewertons) [inline iothub_client/src/iothubtransport_mqtt_common.c:1391]: Nit: should move this free to inside the block where `inputNameCopy` becomes relevant. Otherwise it could be a moot call to free.

Comment (ewertons) [reply iothub_client/src/iothubtransport_mqtt_common.c:1391]: i.e., to the block after malloc succeeds.

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 23]
```diff
@@ -1291,329 +1338,409 @@ static bool isSystemProperty(const char* tokenData)
...
 }
 
+
 //
-// extractMqttProperties parses the MQTT topic PUBLISH'd to this device/module, retrieves properties and fills out the
-// IOTHUB_MESSAGE_HANDLE which will ultimately be delivered to the application callback.
+// findMessagePropertyStart takes an MQTT topic PUBLISH'd as input and returns where message properties, if set, begin.
 //
-static int extractMqttProperties(IOTHUB_MESSAGE_HANDLE IoTHubMessage, const char* topic_name, bool urldecode)
+static const char* findMessagePropertyStart(PMQTTTRANSPORT_HANDLE_DATA transportData, const char* topic_name, IOTHUB_IDENTITY_TYPE type)
```
Comment (ewertons) [inline iothub_client/src/iothubtransport_mqtt_common.c:1532]: Much more readable. Thanks!

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 24]
```diff
@@ -1291,329 +1338,409 @@ static bool isSystemProperty(const char* tokenData)
...
+    }
+
+    return propertiesStart;
+}
+
+//
+// AddApplicationProperty adds the custom key/value property name from the incoming MQTT PUBLISH to the iotHubMessage
+// we will ultimately deliver to the application on its callback.
+//
+static int AddApplicationPropertyToMessage(MAP_HANDLE propertyMap, const char* propertyNameAndValue, size_t propertyNameLength, const char* propertyValue, bool auto_url_encode_decode)
```
Comment (ewertons) [inline iothub_client/src/iothubtransport_mqtt_common.c:1561]: super-nit: You've used camelCase notation in all other static functions, but this one is starting with uppercase.

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 25]
```diff
@@ -1291,329 +1338,409 @@ static bool isSystemProperty(const char* tokenData)
...
+            STRING_delete(propName_decoded);
+        }
+        else if (Map_AddOrUpdate(propertyMap, propertyNameCopy, propertyValue) != MAP_OK)
+        {
+            LogError("Map_AddOrUpdate failed.");
+            result = MU_FAILURE;
+        }
+    }
+
+    free(propertyNameCopy);
```
Comment (ewertons) [inline iothub_client/src/iothubtransport_mqtt_common.c:1608]: This should be after line 1605, in the block after malloc succeeds. Otherwise it's called in case the variable is still null.

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 26]
```diff
@@ -1291,329 +1338,409 @@ static bool isSystemProperty(const char* tokenData)
...
+    return result;
+}
+
+//
+// AddSystemPropertyToMessageWithDecodeIfNeeded adds a "system" property from the incoming MQTT PUBLISH to the iotHubMessage 
+// we will ultimately deliver to the application on its callback.  This function does the urlDecode, if needed, on the property value.
+//
+static int AddSystemPropertyToMessageWithDecodeIfNeeded(IOTHUB_MESSAGE_HANDLE iotHubMessage, MQTT_PROPERTY_TYPE propertyType, const char* propertyValue, bool auto_url_encode_decode)
```
Comment (ewertons) [inline iothub_client/src/iothubtransport_mqtt_common.c:1616]: camelCase notation change?

File: iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c [hunk 27]
```diff
@@ -0,0 +1,1169 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
```
Comment (ewertons) [inline iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c:3]: Could you add a comment in the code here to clarify why we have a second unit test module for iothubtransport_mqtt_common?

**Review by ewertons — APPROVED** [review]
Overall the logic seems correct. Thanks a lot for all this cleanup, John.
